packages/compass-import-export/src/import/guess-filetype.ts

```diff
@@ -0,0 +1,134 @@
+
+function detectJSON(input: Readable): Promise<'json' | 'jsonl' | null> {
+  let jsonVariant: 'json' | 'jsonl' | null = null;
+
+  return new Promise(function (resolve) {
+    const parser = StreamJSON.parser();
+
+    parser.on('data', (data) => {
+      debug('detectJSON:data', data);
+      if (jsonVariant) {
+        return;
+      }
+      if (data.name === 'startObject') {
+        jsonVariant = 'jsonl';
+      } else if (data.name === 'startArray') {
+        jsonVariant = 'json';
+      }
+      parser.destroy();
+    });
+
+    parser.on('end', () => {
+      debug('detectJSON:end', jsonVariant);
+      resolve(jsonVariant);
+    });
+
+    parser.on('close', (err: Error) => {
+      debug('detectJSON:close', err, jsonVariant);
+      resolve(jsonVariant);
+    });
+
+    parser.on('error', (err: Error) => {
+      debug('detectJSON:error', err);
+      resolve(null);
+    });
+
+    input.pipe(parser);
+  });
+}
+
+function hasDelimiterError({ errors }: { errors: { code?: string }[] }) {
+  return (
+    errors.find((error) => error.code === 'UndetectableDelimiter') !== undefined
+  );
+}
+
+function redetectDelimiter({ data }: { data: string[] }): string {
+  for (const char of data[0]) {
+    if (supportedDelimiters.includes(char)) {
+      return char;
+    }
+  }
+
+  return ',';
+}
+
+function detectCSV(input: Readable): Promise<Delimiter | null> {
+  let csvDelimiter: Delimiter | null = null;
+  let lines = 0;
+
+  return new Promise(function (resolve) {
+    Papa.parse(input, {
+      // NOTE: parsing without header: true otherwise the delimiter detection
+      // can't fail and will always detect ,
+      delimitersToGuess: supportedDelimiters as unknown as string[],
+      step: function (results: Papa.ParseStepResult<string[]>, parser) {
+        ++lines;
+        debug('detectCSV:step', lines, results);
+        if (lines === 1) {
+          if (hasDelimiterError(results)) {
+            csvDelimiter = redetectDelimiter(results);
+          } else {
+            csvDelimiter = results.meta.delimiter;
+          }
+        }
+        // must be at least two lines for header row and data
+        if (lines === 2) {
+          parser.abort();
+        }
+      },
+      complete: function () {
+        debug('detectCSV:complete');
+        resolve(lines === 2 ? csvDelimiter : null);
+      },
+      error: function () {
+        debug('detectCSV:error');
+        resolve(null);
+      },
+    });
+  });
+}
+
+type GuessFileTypeOptions = {
+  input: Readable;
+};
+
+type GuessFileTypeResult = {
+  type: 'json' | 'jsonl' | 'csv' | 'unknown';
+  // CSV only
+  csvDelimiter?: Delimiter;
+};
+
+export async function guessFileType({
+  input,
+}: GuessFileTypeOptions): Promise<GuessFileTypeResult> {
+  const jsStream = input.pipe(new PassThrough());
+  const csvStream = input.pipe(new PassThrough());
+
+  const jsonVariant = await detectJSON(jsStream);
+  const csvDelimiter = await detectCSV(csvStream);
```

Comment:
I think you’d need to exhaust the 16kB default streams buffer for this to become an issue, so it’s extremely unlikely to happen, but you could run into an issue here where the JSON parser can’t read any more data from `jsStream` because `input` isn’t piping into it anymore because `csvStream` is not consuming anything.

The easiest way to work around this is probably also the best way, i.e.
```suggestion
const [jsonVariant, csvDelimiter] = await Promise.all([ 
  detectJSON(jsStream), detectCSV(csvStream) 
]);
```